compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1208,19 +1214,33 @@ private void checkRecordVarRefEquivalency(DiagnosticPos pos, BLangRecordVarRef l
             if (variableReference.getKind() == NodeKind.RECORD_VARIABLE_REF) {
                 checkRecordVarRefEquivalency(variableReference.pos,
                         (BLangRecordVarRef) variableReference, rhsField.type, rhsPos);
-            } else if (variableReference.getKind() == NodeKind.TUPLE_VARIABLE_REF) {
+                continue;
+            }
+
+            if (variableReference.getKind() == NodeKind.TUPLE_VARIABLE_REF) {
                 checkTupleVarRefEquivalency(pos, (BLangTupleVarRef) variableReference, rhsField.type, rhsPos);
-            } else if (variableReference.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
+                continue;
+            }
+
+            if (variableReference.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
                 Name varName = names.fromIdNode(((BLangSimpleVarRef) variableReference).variableName);
                 if (varName == Names.IGNORE) {
                     continue;
                 }
-                types.checkType(variableReference.pos,
-                        rhsField.type, variableReference.type, DiagnosticCode.INCOMPATIBLE_TYPES);
-            } else {
-                types.checkType(variableReference.pos,
-                        rhsField.type, variableReference.type, DiagnosticCode.INCOMPATIBLE_TYPES);
             }
+
+            types.checkType(variableReference.pos,
+                    rhsField.type, variableReference.type, DiagnosticCode.INCOMPATIBLE_TYPES);
+        }
+
+        if (lhsVarRef.restParam != null) {
+            boolean hasAnyTypeField = rhsRecordType.fields.stream()
+                    .map(field -> field.type)
+                    .anyMatch(fieldType -> !types.isAnydata(fieldType));
+            types.checkType(((BLangSimpleVarRef) lhsVarRef.restParam).pos, new BMapType(TypeTags.MAP, hasAnyTypeField ?
+                            symTable.anyType : types.isAnydata(rhsRecordType.restFieldType) ?
+                            symTable.anydataType : symTable.anyType, null),
```

Comment:
Shall we move this filter logic to a method for better readability ?

Reply:
Addressed in 417c6e1